File: src/primitives/U256/U256.swift
import Foundation
import CVoltaire

/// 256-bit unsigned integer (big-endian)
public struct U256: Equatable, Hashable, CustomStringConvertible, Sendable {
    internal var raw: PrimitivesU256

    /// Zero value
    public static let zero = U256()

    /// Create zero
    public init() {
        self.raw = PrimitivesU256()
    }

    /// Create from hex string (with or without 0x prefix)
    public init(hex: String) throws {
        var v = PrimitivesU256()
        let rc = hex.withCString { ptr in
            primitives_u256_from_hex(ptr, &v)
        }
        try checkResult(rc)
        self.raw = v
    }

    /// Create from 32 raw bytes (big-endian)
    public init(bytes: [UInt8]) throws {
        guard bytes.count == 32 else {
            throw VoltaireError.invalidLength
        }
        var v = PrimitivesU256()
        withUnsafeMutableBytes(of: &v.bytes) { dest in
            bytes.withUnsafeBytes { src in
                dest.copyMemory(from: src)
            }
        }
        self.raw = v
    }

    /// Convert to hex string (0x-prefixed, 64 hex chars)
    public var hex: String {
        var v = raw
        var buf = [CChar](repeating: 0, count: 67)
        _ = primitives_u256_to_hex(&v, &buf, buf.count)
        return String(cString: buf)
    }

    /// Raw bytes (big-endian)
    public var bytes: [UInt8] {
        withUnsafeBytes(of: raw.bytes) { Array($0) }
    }

    public var description: String { hex }

    public static func == (lhs: U256, rhs: U256) -> Bool {
        lhs.bytes == rhs.bytes
    }

    public func hash(into hasher: inout Hasher) {
        withUnsafeBytes(of: raw.bytes) { hasher.combine(bytes: $0) }
    }
}

File: src/primitives/U256/U256.kt
package com.voltaire

import java.nio.charset.StandardCharsets

/** 256-bit unsigned integer (32 bytes, big-endian) */
class U256 private constructor(private val raw: PrimitivesU256) {

    companion object {
        /** Zero */
        @JvmStatic
        val ZERO: U256 = U256(PrimitivesU256())

        /** Create from hex string */
        @JvmStatic
        @Throws(VoltaireError::class)
        fun fromHex(hex: String): U256 {
            val u = PrimitivesU256()
            checkResult(VoltaireLib.INSTANCE.primitives_u256_from_hex(hex, u))
            return U256(u)
        }

        /** Create from raw bytes (32 bytes, big-endian) */
        @JvmStatic
        @Throws(VoltaireError::class)
        fun fromBytes(bytes: ByteArray): U256 {
            if (bytes.size != 32) throw VoltaireError.InvalidLength
            val u = PrimitivesU256()
            System.arraycopy(bytes, 0, u.bytes, 0, 32)
            return U256(u)
        }

        /** Create from long value */
        @JvmStatic
        fun fromLong(value: Long): U256 {
            val u = PrimitivesU256()
            for (i in 0..7) {
                u.bytes[31 - i] = ((value shr (i * 8)) and 0xFF).toByte()
            }
            return U256(u)
        }
    }

    /** Hex string with 0x prefix */
    val hex: String
        get() {
            val buf = ByteArray(67)
            val rc = VoltaireLib.INSTANCE.primitives_u256_to_hex(raw, buf, 67)
            require(rc >= 0) { "primitives_u256_to_hex failed: $rc" }
            val end = buf.indexOf(0)
            return String(buf, 0, if (end > 0) end else 66, StandardCharsets.UTF_8)
        }

    /** Raw bytes (32 bytes, big-endian) */
    val bytes: ByteArray
        get() = raw.bytes.copyOf()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is U256) return false
        return raw.bytes.contentEquals(other.raw.bytes)
    }

    override fun hashCode(): Int = raw.bytes.contentHashCode()

    override fun toString(): String = hex
}
